### PR TITLE
Add test files

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This is my proposal for the testing setup.

My Idea is that each module gets its own testing. No global test executable
to execute. Instead, we have a common testing helper source (tests.h), which
provides some utilities (by now just a macro for generating the main()
function) and each subdir in `tests/check/` is meant to produce an own
executeable test.

This way, we would have a much better testing setup and one could run only
the appropriate tests while developing.

I consider this setup as really clean. It is _inspired_ by the testing
setup of the gstreamer project, which also uses check.

### CODE REVIEW

File: tests/check/tests.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * @{
+ */
+
+#ifndef __WS_TESTS_TESTS_H__
+#define __WS_TESTS_TESTS_H__
+
+#define WS_TESTS_CHECK_MAIN_ADJUST(genf,ckflags)                   \
+    int                                                            \
+    main(                                                          \
```
Comment (neithernut) [inline tests/check/tests.h]: The `main()` function isn't that long, but I would move the contents to some utility function, taking `genf` and `ckflags` as parameters.

---

Do you actually use `genf`? Can's spot it...

Comment (matthiasbeyer) [inline tests/check/tests.h]: The point of the `main()` function in the macro is that each testing file has it's own `main()` function, which does execute the suite.

---

The call to `genf()` wasn't fixed up after my work, the commit I just added fixes this.

**Comment by matthiasbeyer** [comment]
@neithernut Can you review the build setup carefully? I tried to implement it on my own, I hope I got it right!

**Comment by matthiasbeyer** [comment]
Btw: I'd like to add an additional target for running all of the tests with one command, but I don't know how to do this properly.

I would add a entry in the `.drone.yml` file for it, then. Feel free to do this and send me a quick note from wich of your branches I can pull such changes after you added them (on your own will, of course).

File: tests/check/tests.h (hunk 1 above)
Comment (neithernut) [inline tests/check/tests.h]: Of course each test executable has it's own name, but that's not my point.
Moving the code to a function on it's own keeps the macro short (like, 3 lines), which is always a good thing.

**Comment by neithernut** [comment]
We just do the same we did in `libreset`. We call `ctest`, which in turn runs all the tests.

**Comment by matthiasbeyer** [comment]
So, it seems that this works. @neithernut Is the test setup (the cmake stuff) clean the way I wrote it? Would appreciate your oppinion!

**Comment by neithernut** [comment]
Looks good, you can, however, avoid code repetition in the `CMakeLists.txt` by setting up the targets in a `foreach`. If you want to, I can show you how.

**Comment by matthiasbeyer** [comment]
Feel free to add a commit on top of this branch and notify me for it afterwards (without PR).

**Comment by neithernut** [comment]
That's what I had in mind.

**Comment by neithernut** [comment]
I pushed a branch with a bunch of commits.
Not ethat the test setup still lacks linkage and dependence on the other code.
There are two possibilities to solve this:
- Have a dummy library containing all the symbols we need, constructed from the source files in `src/`
- Add the source files directly

We have to link to all the libraries we need for the source to compile, too.

**Comment by matthiasbeyer** [comment]
Feel free to add more commits.

I'd prefer the direct approach.

**Comment by neithernut** [comment]
You have a GO for squashing.

**Comment by neithernut** [comment]
I'll probably add a last addition anyways.

**Comment by matthiasbeyer** [comment]
Done.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**